docs/contributing/inter-thread-context-propagation.md

```diff
@@ -0,0 +1,94 @@
+# The story of context propagation across threads
+
+## The need
+Take a look at the following two pseudo-code snippets (see below for explanations).
+
+```
+Executor pool = Executors.newFixedThreadPool(10)
+
+public void doGet(HttpServletRequest request, HttpServletResponse response) {
+    Future f1 = pool.submit(() -> {
+        return userRepository.queryShippingAddress(requet)
+    })
+    Future f2 = pool.submit(() -> {
+        return warehouse.currentState(requet)
+    })
+    writeResponse(response, f1.get(), f2.get())
+}
+```
+
+```
+Executor pool = Executors.newFixedThreadPool(10)
+
+public void doGet(HttpServletRequest request, HttpServletResponse response) {
+    final AsyncContext acontext = request.startAsync();
+    acontext.start(() -> {
+            String address = userRepository.queryShippingAddress(requet)
+            HttpServletResponse response = acontext.getResponse();
+            writeResponse(response, address)
+            acontext.complete();
+   }
+}
+```
+
+In both cases request processing requires some potentially long operation and application developer
+wants to do them off the main thread. In the first case this hand-off between request accepting thread
+and request processing thread happens manually, by submitting work into some thread pool.
+In the second case it is the framework that handles separate thread pool and passing work to it.
+
+In cases like this proper tracing solution should still combine into a single trace all the work
+required for request processing, regardless in what thread that work happened. With proper
+parent-child relationship between span: span representing shipping address query should be the child
+of the span which denotes accepting HTTP request.
+
+## The solution
+Java auto instrumentation uses an obvious solution to the requirement above: we attach current execution
+context (represented in the code by `io.grpc.Context`) with each `Runnable`, `Callable` and `ForkJoinTask`.
+"Current" means the context active on the thread which calls `Executor.execute` (and its analogues
+such as `submit`, `invokeAll` etc) at the moment of that call. Whenever some other thread starts
+actual execution of that `Runnable` (or `Callable` or `ForkJoinTask`), that context get restored
+on that thread for the duration of the execution. This can be illustrated by the following pseudo-code:
+
+```
+    var job = () -> {
+        try(Scope scope = withScopedContext(this.context)) {
+            return userRepository.queryShippingAddress(requet)
+        }}
+    job.context = Context.current()
+    Future f1 = pool.submit()
+
+```
+
+## The drawback
+There are some runtime environments which, simplifying, do the following:
+```
+pool.submit(new AcceptRequestRunnable() {
+    Request req = readRequest()
+    // At this point Context.current() will have a started span
```

Comment:
> At this point Context.current() will have a started span

this is the part of the example that I'm not following, how does the (thread-bound) `Context` with the started span escape `readRequest()`?

Reply:
And now?